remote_file_system/server.py

```diff
@@ -46,28 +46,37 @@ def __init__(
         # store request id as key. value is the Message
         self.message_history: Dict[UUID, Message] = {}
         self.invocation_semantics = invocation_semantics
+        self.keep_listening = True
+
+    def stop_listening(self) -> None:
+        self.keep_listening = False
 
     def listen_for_messages(self) -> None:
         sock = socket(AF_INET, SOCK_DGRAM)
         server_address: Tuple[str, int] = (str(self.server_ip_address), self.server_port_number)
         sock.bind(server_address)
 
+        sock.settimeout(10)
+        # sock.setblocking(False)
+
         try:
-            while True:
-                logger.info(f"Socket is listening for messages at {self.server_ip_address}:{self.server_port_number}.")
-
-                incoming_bytes, sender_address = sock.recvfrom(4096)  # TODO review fixed buffer size
-                sender_ip_address, sender_port_number = sender_address
-                logger.info(f"Received {len(incoming_bytes)} bytes from {sender_ip_address}:{sender_port_number}.")
-
-                if incoming_bytes:
-                    incoming_message: Message = Message.unmarshall(incoming_bytes)
-                    logger.debug(f"Received {incoming_message}.")
-                    self._dispatch_message(
-                        message=incoming_message,
-                        client_ip_address=IPv4Address(sender_ip_address),
-                        client_port_number=sender_port_number,
-                    )
+            try:
+                while self.keep_listening:
+                    logger.info(f"Socket is listening for messages at {self.server_ip_address}:{self.server_port_number}.")
+                    incoming_bytes, sender_address = sock.recvfrom(4096)  # TODO review fixed buffer size
+                    sender_ip_address, sender_port_number = sender_address
+                    logger.info(f"Received {len(incoming_bytes)} bytes from {sender_ip_address}:{sender_port_number}.")
+
+                    if incoming_bytes:
+                        incoming_message: Message = Message.unmarshall(incoming_bytes)
+                        logger.debug(f"Received {incoming_message}.")
+                        self._dispatch_message(
+                            message=incoming_message,
+                            client_ip_address=IPv4Address(sender_ip_address),
+                            client_port_number=sender_port_number,
+                        )
+            except TimeoutError:
+                pass
         finally:
             sock.close()
```

Comment:
I don't think we need nested `try` blocks

Reply:
The nested try block is for the TimeoutError. Because I used the sock.settimeout(10), this will invoke a timeouterror if the server did not receive any messages after 10 seconds. This done so that we can go into the next iteration of the while loop.